File: go/0036.go
package leetcode

func isValidSudoku(board [][]byte) bool {
	col_checks := [9][9]byte{}
	block_checks := [9][9]byte{}
	for row_idx, row := range board {
		row_check := [9]byte{}
		for col_idx, cell := range row {
			if string(cell) == "." {
				continue
			}
			cell_num := int(cell - '1')
			if row_check[cell_num] > 0 {
				return false
			}
			if col_checks[col_idx][cell_num] > 0 {
				return false
			}
			block_idx := row_idx/3*3 + col_idx/3
			if block_checks[block_idx][cell_num] > 0 {
				return false
			}

			row_check[cell_num] = 1

			col_checks[col_idx][cell_num] = 1

			block_checks[block_idx][cell_num] = 1
		}
	}
	return true
}

File: python/0036.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        col_nums = [set() for i in range(9)]
        box_nums = [set() for i in range(9)]
    
        for row_idx, row in enumerate(board):
            row_nums = set()
            for col_idx, cell in enumerate(row):
                box_idx = int(col_idx / 3) + 3 * int(row_idx / 3)
                if cell == ".":
                    continue
                num = int(cell)
                if num in col_nums[col_idx] or num in row_nums or num in box_nums[box_idx]:
                    return False
                col_nums[col_idx].add(num)
                row_nums.add(num)
                box_nums[box_idx].add(num)

                

        return True

        
        
        